src/partition/src/multi_dim.rs

```diff
@@ -0,0 +1,201 @@
+}
+
+impl MultiDimPartitionRule {
+    pub fn new(
+        partition_columns: Vec<String>,
+        regions: Vec<RegionNumber>,
+        exprs: Vec<PartitionExpr>,
+    ) -> Self {
+        let name_to_index = partition_columns
+            .iter()
+            .enumerate()
+            .map(|(i, name)| (name.clone(), i))
+            .collect::<HashMap<_, _>>();
+
+        Self {
+            partition_columns,
+            name_to_index,
+            regions,
+            exprs,
+        }
+    }
+
+    fn find_region(&self, values: &[Value]) -> Result<RegionNumber> {
+        ensure!(
+            values.len() == self.partition_columns.len(),
+            error::RegionKeysSizeSnafu {
+                expect: self.partition_columns.len(),
+                actual: values.len(),
+            }
+        );
+
+        for (region_index, expr) in self.exprs.iter().enumerate() {
+            if self.evaluate_expr(expr, values)? {
+                return Ok(self.regions[region_index]);
+            }
+        }
+
+        // return the default region number
+        Ok(0)
+    }
+
+    fn evaluate_expr(&self, expr: &PartitionExpr, values: &[Value]) -> Result<bool> {
+        match (expr.lhs.as_ref(), expr.rhs.as_ref()) {
+            (Operand::Column(name), Operand::Value(r)) => {
+                let index = self.name_to_index.get(name).unwrap();
+                let l = &values[*index];
+                Self::perform_op(l, &expr.op, r)
+            }
+            (Operand::Value(l), Operand::Column(name)) => {
+                let index = self.name_to_index.get(name).unwrap();
+                let r = &values[*index];
+                Self::perform_op(l, &expr.op, r)
+            }
+            (Operand::Expr(lhs), Operand::Expr(rhs)) => {
+                let lhs = self.evaluate_expr(lhs, values)?;
+                let rhs = self.evaluate_expr(rhs, values)?;
+                match expr.op {
+                    RestrictedOp::And => Ok(lhs && rhs),
+                    RestrictedOp::Or => Ok(lhs || rhs),
+                    _ => unreachable!(),
+                }
+            }
+            _ => unreachable!(),
+        }
+    }
+
+    fn perform_op(lhs: &Value, op: &RestrictedOp, rhs: &Value) -> Result<bool> {
+        let result = match op {
+            RestrictedOp::Eq => lhs.eq(rhs),
+            RestrictedOp::NotEq => lhs.ne(rhs),
+            RestrictedOp::Lt => lhs.partial_cmp(rhs) == Some(Ordering::Less),
+            RestrictedOp::LtEq => {
+                let result = lhs.partial_cmp(rhs);
+                result == Some(Ordering::Less) || result == Some(Ordering::Equal)
+            }
+            RestrictedOp::Gt => lhs.partial_cmp(rhs) == Some(Ordering::Greater),
+            RestrictedOp::GtEq => {
+                let result = lhs.partial_cmp(rhs);
+                result == Some(Ordering::Greater) || result == Some(Ordering::Equal)
+            }
+            RestrictedOp::And | RestrictedOp::Or => unreachable!(),
```

Comment:
Are they unsupported or unreachable?

Reply:
Unreachable. Conjunction is handled by caller

Reply:
ok, let's add a comment.